docker-compose-rule-core/src/main/java/com/palantir/docker/compose/execution/DockerCommandLocations.java

```diff
@@ -1,47 +1,75 @@
 /*
  * Copyright 2016 Palantir Technologies, Inc. All rights reserved.
- *
- * Licensed under the Apache License, Version 2.0 (the "License");
- * you may not use this file except in compliance with the License.
- * You may obtain a copy of the License at
- *
- *     http://www.apache.org/licenses/LICENSE-2.0
- *
- * Unless required by applicable law or agreed to in writing, software
- * distributed under the License is distributed on an "AS IS" BASIS,
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
- * limitations under the License.
  */
-package com.palantir.docker.compose.execution;
 
-import static java.util.Arrays.asList;
+package com.palantir.docker.compose.execution;
 
 import java.io.File;
-import java.util.List;
-import java.util.Optional;
-import java.util.function.Predicate;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Map;
+import java.util.regex.Pattern;
+import java.util.stream.Stream;
+import javax.annotation.Nullable;
+import org.immutables.value.Value;
 
-public class DockerCommandLocations {
-    private static final Predicate<String> IS_NOT_NULL = path -> path != null;
-    private static final Predicate<String> FILE_EXISTS = path -> new File(path).exists();
+@Value.Immutable
+public abstract class DockerCommandLocations {
 
-    private final List<String> possiblePaths;
+    private static final Pattern PATH_SPLITTER = Pattern.compile(File.pathSeparator);
 
-    public DockerCommandLocations(String... possiblePaths) {
-        this.possiblePaths = asList(possiblePaths);
+    @Value.Default
+    protected Map<String, String> env() {
+        return System.getenv();
     }
 
-    public Optional<String> preferredLocation() {
+    @Value.Derived
+    protected String path() {
```

Comment:
This doesn't work on Windows, where the path variable is "Path". You could add another `if (path == null) path = env().get("Path")`, or iterate through the env() map to find any key that equals "path" ignoring case, e.g.

```
        String path = getEnv().entrySet().stream()
                .filter(e -> e.getKey().equalsIgnoreCase("path"))
                .findFirst()
                .map(Map.Entry::getValue)
                .orElseThrow(() -> new IllegalStateException("Could not find path variable in env"));
```